src/contracts/EthBridge.h

```diff
@@ -0,0 +1,634 @@
+                CONTRACT_INDEX,
+                EthBridgeError::onlyManagersCanCompleteOrders, 
+                0, // No order ID involved
+                0, // No amount involved
+                '\0'
+            };
+            LOG_INFO(locals.log);
+
+            output.status = 1; // Error
+            return;
+        }
+
+        state.managers.removeByKey(input.address); // Remove manager
+        locals.log = EthBridgeLogger{
+            CONTRACT_INDEX,
+            0, // No error
+            0, // No order ID involved
+            0, // No amount involved
+            '\0'
+        };
+        LOG_INFO(locals.log);
+        output.status = 0; // Success
+    _
+
+    struct getTotalReceivedTokens_locals {
+        EthBridgeLogger log;
+    };
+
+    PUBLIC_FUNCTION_WITH_LOCALS(getTotalReceivedTokens)
+
+        locals.log = EthBridgeLogger{
+            CONTRACT_INDEX,
+            0, // No error
+            0, // No order ID involved
+            state.totalReceivedTokens, // Amount of total tokens
+            '\0'
+        };
+        LOG_INFO(locals.log);
+        output.totalTokens = state.totalReceivedTokens;
+    _
+
+    struct completeOrder_locals {
+        EthBridgeLogger log;
+    };
+
+    // Complete an order and release tokens
+    PUBLIC_PROCEDURE_WITH_LOCALS(completeOrder)
+        isManager_input invocatorAddress = qpi.invocator();
+        isManager_output isManagerOperating = false;
+        CALL(isManager, invocatorAddress, isManagerOperating);
+
+        if (!isManagerOperating) {
+            locals.log = EthBridgeLogger{
+                CONTRACT_INDEX,
+                EthBridgeError::onlyManagersCanCompleteOrders,
+                input.orderId,
+                0,
+                '\0'
+            };
+            LOG_INFO(locals.log);
+            output.status = 1; // Error
+            return;
+        }
+
+        // Retrieve the order
+        BridgeOrder order;
+        if (!state.orders.get(input.orderId, order)) {
+            locals.log = EthBridgeLogger{
+                CONTRACT_INDEX,
+                EthBridgeError::invalidOrderState,
+                input.orderId,
+                0,
+                '\0'
+            };
+            LOG_INFO(locals.log);
+            output.status = 1; // Error
+            return;
+        }
+
+        // Check the status
+        if (order.status != 0) { // Ensure it's not already completed or refunded
+            locals.log = EthBridgeLogger{
+                CONTRACT_INDEX,
+                EthBridgeError::invalidOrderState,
+                input.orderId,
+                0,
+                '\0'
+            };
+            LOG_INFO(locals.log);
+            output.status = 2; // Error
+            return;
+        }
+
+        // Handle order based on transfer direction
+        if (order.fromQubicToEthereum) {
+            // Ensure sufficient tokens were transferred to the contract
+            if (state.totalReceivedTokens < order.amount) {
```

Comment:
should this be `lockedTokens`? Because I don't see `totalReceivedTokens` get decreased anywhere in this contract, meaning it's only increasing

Reply:
Absolutely true, that's my fault! I modified the logic of completeOrdder, refundOrder and transferToContract so now makes more sense. 

`totalReceivedTokens`: tracks the total of all tokens sent to the contract, that are not currently in use (locked). It reflects the overall balance the contract has in tokens that are not actively tied to any order.

`lockedTokens`: represents the amount of the tokens that are actively locked in pending or active orders. It ensures that the contract knows how many tokens are reserved and unavailable for other operations.

Now it follows this **logic**:
- When a user transfers tokens to the contract (via transferToContract), the `totalReceivedTokens` increases. These tokens are not yet locked until they are associated with an order. This way, the contract keeps track of unused tokens.
- When an order is in the completion process, the required tokens are deducted from the available `totalReceivedTokens` (totalReceivedTokens -= input.amount) and added to `lockedTokens` (lockedTokens += input.amount).
- If an order is refunded or completed:
`lockedTokens` is decreased because the tokens are no longer reserved.
`totalReceivedTokens` remains unchanged, as it no longer was considering those 'lockedTokens' in its balance.